### ISSUE TITLE
Equals not found when using lazy view

### ISSUE THREAD

**Issue by forki** [fable-compiler/fable-elmish#17]
I tried to use lazyviews: https://github.com/fable-compiler/fable-react_native-demo/blob/lazyview/src/Nightwatch.fs#L24

but it can't find Equals:

![image](https://cloud.githubusercontent.com/assets/57396/20426536/d5423ab6-ad7f-11e6-8b8b-b25eab381810.png)

![image](https://cloud.githubusercontent.com/assets/57396/20426710/a6c6f072-ad80-11e6-938d-c488e58c75a9.png)

The AppModel has a Equals function as you can see in App.js

/cc @et1975 @alfonsogarciacaro

**Comment by alfonsogarciacaro** [comment]
Maybe this is related to what we talked before of React stripping off the prototype methods of the state object? Could this be solved by using `toPlainJsObject` when setting the state too? Or maybe we need to create some helpers to handle the state of React components that use Fable equality methods...

**Comment by forki** [comment]
but @et1975 is using this on Model classes. so how can this be a problem?

**Comment by forki** [comment]
looks like https://github.com/fable-compiler/fable-react_native-demo/commit/9137e7999179f0fc01f05c77b118e3f8d3aa69bb is indeed working. weird.

**Comment by alfonsogarciacaro** [comment]
At the end we'll probably have to define our components to deal with equality of props and state.

**Comment by forki** [comment]
In this case I think I actually want to handle it per subview. Still weird behavior

**Comment by et1975** [comment]
@forki Yes, I've seen the problem. 
There are a few possibilities:
- as @alfonsogarciacaro mentioned, prototype methods like `Equals` might have been stripped -React batches some updates which results in a new JS object, unrelated to the prototype of the original, passed as the `state`. `LazyView` uses props, which are already constructed using `toPlainJsObject`, and the root state object I'm converting in the `App` component. It shouldn't be an issue, but I'm not 100% certain.
- fable-core@0.6.x does deep comparison (doesn't shortcut on reference equality) which may lead to some unexpected object in the tree being compared (and therefor requiring `equality`). If you've got some deserialized JSON (like webservice response) as part of the state tree it might endup being compared and needs to implement the equality.

TL;DR: For now you may have to supply a customer comparer into `lazyView2`: `lazyView2With (fun x y -> obj.ReferenceEquals(x,y))` or something like that. Redefine 'lazyView2` in your Prelude with that, until elmish can target fable-core@0.7.

**Comment by alfonsogarciacaro** [comment]
Although so far I've been a bit reluctant to it, maybe it's time to introduce a `[<PlainObj>]` (or POJO) record that compiles directly to a JS object and for which the compiler can makes some restriction (no static members, structural equality...). We could also add attributes to arguments to make sure the passed type is a plain object

What do you think, @forki, @et1975? Would it help interaction with React?

**Comment by et1975** [comment]
I don't think it's necessary (for React compat). With regards to equality, there are things that elmish can do and there are things that neither fable, nor elmish can help with - for example embedding JS objects into the state, such as RN navigation state which is constructed by RN and is never going to have a comparer will require the end user to know that and supply the custom impl.

**Comment by et1975** [comment]
Apparently I forgot to publish the fix for the root component to support lazy view (of the root model), I tweaked it a bit and it's now released as fable-elmish-react@0.2.2.

One gotcha I want to mention as we've hit internally, is that when a `view` is being made `lazy` the model being compared has to include the navigation state (if any). The example of the opposite would be something like:
```fsharp
fun render model dispatch =
     match model.navigationState with
     | Scene1 -> lazyView2 scene1render model.scene1 (Scene1 >> dispatch)
     | Scene2 -> lazyView2 scene2render model.scene2 (Scene2 >> dispatch)
```

the problem with the above is that the `model.navigationState` doesn't partake in the equality check of the lazy children. The correct version would look something like this:

```fsharp
fun render model dispatch =
     match model.navigationState with
     | Scene1 -> lazyView3 (fun _ -> scene1render) model.navigationState model.scene1 (Scene1 >> dispatch)
     | Scene2 -> lazyView3 (fun _ -> scene2render) model.navigationState model.scene2 (Scene2 >> dispatch)
```

There are different ways to write this, but the gist of it is - navigation state affects the lazy content, but it doesn't force any changes in the signatures of children.

**Comment by et1975** [comment]
@forki can you confirm that fable-elmish-react@0.2.2 fixes it?

**Comment by forki** [comment]
yes can't reproduce any more. thx
